pdfwhodunit.1
.TH pdfwhodunit 1 "July 21, 2019"

.
.
.
.SH NAME
pdfwhodunit, pdftrace.so - trace a PDF file

.
.
.
.SH SYNOPSIS
[LD_PRELOAD=pdftrace.so] [GRANULARITY=n] \fBpdfwhodunit\fP file.pdf [box]

LD_PRELOAD=\fBpdftrace.so\fP [GRANULARITY=n] program file.pdf

.
.
.
.SH DESCRIPTION

\fBpdfwhodunit\fP tells which parts of a PDF file hold the content of an area
of the page. It requires the \fBpdftrace.so\fP preload library, which must be
in the library path (\fILD_LIBRARY_PATH\fP or \fI/etc/ld.so.conf\fP), in the
current directory or explicitely specified via the \fILD_PRELOAD\fP variable.
The \fBpdftrace.so\fP preload library can be used with other programs to trace
their access to PDF files.

.SS pdfwhodunit

If the PDF file contains something in the box passed as argument,
\fBpdfwhodunit\fP traces it back to its position in the PDF file. An example
call and output is:

.nf
> pdfwhodunit file.pdf [100,120-200,300]
\&...
read from 3913      to 4169
OBJECT 5

read from 9801      to 10057
OBJECT 5
.fi

The program is requested to trace the rectangular area from the point
\fI100,120\fP to the point \fI200,300\fP in the file \fIfile.pdf\fP. It answers
that the given rectangle contains some text, lines or raster data originating
between location \fI3913\fP and \fI4169\fP in the file and between \fI9801\fP
and \fI10057\fP. These two zones of the file contain what is drawn in the box,
more or less (tracing is not exact: see below for details).

Normally, the content of PDF files is encoded and therefore impossible to read
and change by hand. Recoding with \fIqpdf(1)\fP with option \fI-qdf\fP turns
most parts of it into text:

.nf
.I qpdf -qdf input.pdf output.pdf
.fi

The box is either \fI[x1,y1,x2,y2]\fP or \fI[x,y+width,height]\fP and defaults
to the whole page. It can be determined for example in \fIhovacui(1)\fP:

.IP "  * " 4
switch to "fit none" mode by key 'f' or commandline argument \fI-fn\fP
.IP "  * "
zoom to the part of the page to trace
.IP "  * "
save the coordinates of the area currently visible by key 'B'
.IP "  * "
retrieve the box from \fIhovacui-out.txt\fP

.P

If the last line of \fIhovacui-out.txt\fP is for example
\fI[120,100-320,180]\fP, the content of that box is located in the PDF file can
with \fPpdfwhodunit file.pdf [120,100-320,180]\fP. If it begins at position
\fI1234\fP in the file, it can be shown in \fIvim(1)\fP by the command
\fI:1234go\fP.

Technically, \fIpdfwhodunit\fP only traces file reads. The displayed intervals
\fI3913-4169\fP and \fI9801-10057\fP are the last zones read from
\fIfile.pdf\fP before something is drawn in the rectangle
\fI[100,120-200,300]\fP. These may not be exactly the parts of the file that
hold the content of the box.

\fIpdfwhodunit\fP reads the PDF file and renders it on an internal (invisible)
buffer. Whenever something changes in the box that is passed as argument, it
prints which portion of the file was last read. The content of that portion is
what caused the box to change content. Also printed is the number of PDF object
the file portion belongs to.

The actual graphical operations (e.g., print a string, draw a line) that
changed the content of the box may be in a previous portion since they only
take effect as a result of a stroke operation, which may appear later in the
file.

The size of the portions of the PDF file as printed by \fIpdfwhodunit\fP is
decided by the \fIpoppler(3)\fP library, and is typically 256 bytes. It can be
reduced by the environment variable \fIGRANULARITY\fP. A small value makes
tracing more precise but slower. A too small value makes \fIpdfwhodunit\fP
unable to identify the PDF objects, which are therefore not printed.

.SS pdftrace.so

\fIpdfwhodunit\fP works by tracing the accesses to the PDF by the
\fIpoppler(3)\fP library. This is done by intercepting the calls to
\fIopen(2)\fP, \fIread(2)\fP and similar before they reach the standard C
library \fIlibc.so\fP. The mechanism that makes this possible is the
\fILD_PRELOAD\fP library \fIpdftrace.so\fP.

The same mechanism can be used with other programs as well. For example, one
may trace a PDF file while it is displayed:

.nf
LD_PRELOAD=pdftrace.so some_pdf_viewer file.pdf
.fi

In this case, \fIpdftrace.so\fP stops every time the program
\fIsome_pdf_viewer\fP reads from the file. It continues when either the
\fIF3\fP or the \fIF4\fP key is pressed. The first makes it proceed until the
next read, the second until reading switches to another PDF object. If the PDF
viewer displays content right after reading it from the file, this mechanism
shows which parts of the file hold which page content.

The same limitations described above apply: this is done by file reads, so the
actual content may be located before the point that causes it to be displayed.
How much to read each time is decided by the \fIpoppler(3)\fP library (but can
be changed by the GRANULARITY environment variable). Additionally, the
mechanism of stopping and printing is only useful if the content of the file is
displayed as soon as it is read. It does not make sense if some form of double
buffering is employed.

.
.
.SH EXAMPLE

The article \fIthe-theory-of-nothing.pdf\fP contains the obvious error that
your name \fIProf. Dr. Franz Henrich Von Hausseldorfstain-Roberts Gonzalez\fP
is spelled wrong, as it misses an 'n'. You do not have the LaTeX source written
by your graduate students and are in a hurry to have it published. Being a
well-educated person, you know that PDF files can be edited, contrary to the
popular belief they cannot. The problem is that the file looks mostly like
garbage when opened in a text editor, something like:

.nf
HaxB5mGJq+n#"bGr5@Wc%S6F%-_((rZJ(Ws]ZQq)QIXQRGX"E&>bN%;'8.W*+q7>*
&_B$g3kLiQEA!uM_hL"rUKW1E.,=pkOa[:b)dg4W-sO_@%4O>K<0XEAdfA0bnd48S
2HPkCkrf=64PWp5BC`$Jc:X.OU!5;C=?9!da#0H7&W6e:BWq[A0FG4<O24BVVcH;;
j*r)/"<XLr#!1ZC$m%K>r4Q:!ACadUPA`Q'1-1F"<NM8$n&KK$I&mk%3mZDF1c/=^
_"EpZ"WIKY;?MRTW5c1uY]CQsD#Ue4gscRLNK.86pt=81_u"Q&9rb,b7ond_1Qb.g
Wiq%%1_LWkP9#94%h-mh2Lj$;L4#tHfu/u!7kkr`RLPdG8j%Be0:OAnMe1@:A^oc]
.fi

The first step is to make it readable by \fIqpdf(1)\fP:

.nf
qpdf -qdf the-theory-of-nothing.pdf readable.pdf
.fi

The second step is to locate your name in the file. Still better, locate the
position of the missing 'n'. You open the file with \fIhovacui(1)\fP in no-fit
mode:

.nf
hovacui -fn readable.pdf
.fi

Using 'z' to zoom and the cursor keys to move, you position the PDF viewer to
the place of the missing letter in "Herich", making this word large enough so
that nothing else is in the screen.

Key 'B' tells the position of the corners of the screen in the document. You
forget to write them down, but you do not worry: after closing the viewer, you
find them in \fIhovacui-out.txt\fP. If the last line of this file is for
example \fI[112.942,130.351-115.430,135.543]\fP, you pass that to
\fIpdfwhodunit\fP:

.nf
GRANULARITY=10 pdfwhodunit readable.pdf [112.942,130.351-115.430,135.543]
.fi

The output contains:

.nf
\&...
read from 4512      to 4522
OBJECT 10
.fi

Now you open the pdf file with:

.nf
vim readable.pdf
.fi

And move to the position of interest with \fI:4512go\fP. Around this location,
the file contains:

.nf
(Prof. ) Tj
(Franz He) Tj
(rich Von Hausse) Tj
.fi

You insert the missing 'n'. The file looks good except for an annoying warning
about the xref blah blah blah. You fix it by again qpdf:

.nf
qpdf readable.pdf the-theory-of-nothing-fixed.pdf
.fi

And now the article is ready to be published!



.
.
.
.SH STRUCTURE OF PDF

A PDF file comprises a number of object, numbered from 1 on. Each object begins
with an header line that specifies its number, like \fI21 0 obj\fP. It ends
with the trailer line \fIendobj\fP. Example objects are:

.IP "  * " 4
the list of the pages in the document, an object number for each
.IP "  * "
the object for a page, with the number of the object holding its content
.IP "  * "
the object that holds the content of the page
.IP "  * "
the object for a font
.IP "  * "
the object for a raster image

.P
The content of a page is given by commands like:

.nf
BT                             % begin a text block
/FONT1 8 Tf                    % select a font
20 40 Td                       % move cursor to position 20,40
(abcd) Tj                      % write the string "abcd"
ET                             % end a text block
20 80 60 -60 re                % draw a rectangle
S                              % stroke
.fi

\fIpdfwhodunit\fP knows nothing of this. It simply presumes that if something
is drawn in the box passed as argument right after reading a certain block of
the file, that block contains what is in the box. But the actual box content
may be earlier in the file. For example, the rectangle \fI20 80 60 -60 re\fP is
actually drawn only as a result of the stroke \fIS\fP, which could have been
much later in the file.

.
.
.
.SH ENVIRONMENT VARIABLES

.TP
.B
LD_PRELOAD
set to \fIpdftrace.so\fP to trace PDF file access by an arbitrary program; this
is done automatically by \fIpdfwhodunit\fP; depending on the permissions, the
value of this variable might be just the library file name, an absolute or
relative path
.TP
.B
GRANULARITY
trace the input file in blocks of this size, by default 256; small numbers
increase the precision of tracing, but decrease efficiency; when it is too
small, object starts are not detected
.TP
.B
DOUBLEBUFFERING
\fIpdftrace.so\fP sets it to "no" unless already set in the environment



pdfwhodunit.c
/*
 * pdfwhodunit.c
 *
 * trace a PDF file
 *
 * see pdfwhodunit.1
 * also comments at the top of pdftrace.c
 */

#define _GNU_SOURCE
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <dlfcn.h>

#include <poppler.h>
#include <cairo.h>
#include <cairo-pdf.h>

#define HOMELIB "bin"

/*
 * current and previous images
 */
cairo_surface_t *surface;
int size;
unsigned char *image;
unsigned char *previous;

/*
 * tracer
 */
int tracer() {
	cairo_surface_flush(surface);
	if (! memcmp(previous, image, size))
		return 0;
	memcpy(previous, image, size);
	return 1;
}

/*
 * main
 */
int main(int argc, char *argv[]) {
	char *preload, *home, *var;
	int opt;
	gboolean usage = FALSE;
	char *infile;
	gdouble rwidth, rheight;
	char c, d;

	void (*pdftrace)();
	void (*usenoterminal)();
	void (*register_tracer)(int (*t)());
	void (*traced_final)();

	GFile *gin;
	GError *err = NULL;
	int format = CAIRO_FORMAT_ARGB32;
	PopplerDocument *doc;
	PopplerPage *page;
	PopplerRectangle clip = {0.0, 0.0, 0.0, 0.0};
	int npages, n;
	gdouble width = 200, height = 200, stride;
	gdouble pwidth, pheight;
	cairo_t *cr;

				/* probe pdftrace.so */

	pdftrace = dlsym(RTLD_DEFAULT, "pdftrace");
	if (pdftrace == NULL) {
		printf("%s\n", dlerror());
		printf("not called with pdftrace.so preloaded\n");

		preload = getenv("LD_PRELOAD");
		if (preload == NULL) {
			printf("trying LD_PRELOAD=./pdftrace.so\n");
			putenv("LD_PRELOAD=./pdftrace.so");
			execvp(argv[0], argv);
			perror(argv[0]);
			exit(EXIT_FAILURE);
		}
		else if (! strcmp(preload, "./pdftrace.so")) {
			printf("trying LD_PRELOAD=pdftrace.so\n");
			putenv("LD_PRELOAD=pdftrace.so");
			execvp(argv[0], argv);
			perror(argv[0]);
			exit(EXIT_FAILURE);
		}
		else if (! strcmp(preload, "pdftrace.so")) {
			home = getenv("HOME");
			if (home != NULL) {
				var = malloc(strlen(home) + 100);
				sprintf(var, "LD_PRELOAD=%s/%s",
					home, "/" HOMELIB "/pdftrace.so");
				putenv(var);
				execvp(argv[0], argv);
				perror(argv[0]);
				exit(EXIT_FAILURE);
			}
		}

		printf("cannot find pdftrace.so\n");
		printf("call as:\n");
		printf("LD_PRELOAD=pdftrace.so pdfwhodunit ...\n");
		printf("or\n");
		printf("LD_PRELOAD=./pdftrace.so pdfwhodunit ...\n");
		exit(EXIT_FAILURE);
	}

				/* arguments */

	while ((opt = getopt(argc, argv, "h")) != -1)
		switch(opt) {
		case 'h':
			usage = TRUE;
			break;
		}

	if (! usage && argc - 1 < optind) {
		printf("input file name missing\n");
		usage = TRUE;
	}
	if (usage) {
		printf("usage:\n");
		printf("\tLD_PRELOAD=./pdftrace.so [GRANULARITY=n] \\\n");
		printf("\tpdfwhodunit ");
		printf("[-h] file.pdf [x1,y1,x2,y2]\n");
		printf("\t\t-h\t\tthis help\n");
		exit(EXIT_FAILURE);
	}
	infile = argv[optind];
	if (argc - 1 > optind) {
		if (5 == sscanf(argv[optind + 1], "[%lg,%lg+%lg,%lg%c%c",
				&clip.x1, &clip.y1, &rwidth, &rheight,
				&c, &d) && c == ']') {
			clip.x2 = clip.x1 + rwidth;
			clip.y2 = clip.y1 + rheight;
		}
		else if (5 == sscanf(argv[optind + 1], "[%lg,%lg-%lg,%lg%c%c",
				&clip.x1, &clip.y1, &clip.x2, &clip.y2,
				&c, &d) && c == ']') {
		}
		else {
			printf("error parsing rectangle: ");
			printf("%s\n", argv[optind + 1]);
			exit(EXIT_FAILURE);
		}
		printf("clipping rectangle: ");
		printf("[%g,%g,%g,%g]\n", clip.x1, clip.y1, clip.x2, clip.y2);
	}

				/* do not use terminal codes */

	usenoterminal = dlsym(RTLD_DEFAULT, "usenoterminal");
	usenoterminal();

				/* create images */

	stride = cairo_format_stride_for_width(format, width);
	size = stride * height;
	image = malloc(size);
	memset(image, 0, size);
	previous = malloc(size);
	memcpy(previous, image, size);
	surface = cairo_image_surface_create_for_data(image,
		format, width, height, stride);

				/* register tracer */

	register_tracer = dlsym(RTLD_DEFAULT, "register_tracer");
	register_tracer(tracer);

				/* open file */

	gin = g_file_new_for_path(infile);
	doc = poppler_document_new_from_gfile(gin, NULL, NULL, &err);
	if (doc == NULL) {
		g_printerr("error opening pdf file: %s\n", err->message);
		g_error_free(err);
		exit(EXIT_FAILURE);
	}

				/* pages */

	npages = poppler_document_get_n_pages(doc);
	if (npages < 1) {
		printf("no page in document\n");
		exit(EXIT_FAILURE);
	}

				/* copy to destination */

	printf("pages: \n");

	for (n = 0; n < npages; n++) {
		printf("  - page: %d\n", n);
		page = poppler_document_get_page(doc, n);
		poppler_page_get_size(page, &pwidth, &pheight);

		cr = cairo_create(surface);
		cairo_scale(cr, width / pwidth, height / pheight);
		if (clip.x2 != 0 || clip.y2 != 0) {
			cairo_rectangle(cr, clip.x1, clip.y1,
				clip.x2 - clip.x1, clip.y2 - clip.y1);
			cairo_clip(cr);
		}
		poppler_page_render_for_printing(page, cr);
		cairo_destroy(cr);
		cairo_surface_show_page(surface);

		g_object_unref(page);
	}

	traced_final = dlsym(RTLD_DEFAULT, "traced_final");
	traced_final();

	cairo_surface_destroy(surface);
	g_object_unref(gin);

	return EXIT_SUCCESS;
}

